### PR TITLE
[ntop] add support for arrays and IPv6

### PR DESCRIPTION
**Pull request by mmarchini**
- Added ntop support for array types of sizes 4 and 16
- Added ntop support for IPv6 
- Optimized number of instructions generated for ntop and memory used, both in the BPF stack and maps
- Make first argument of ntop optional, inferring the af_type from the address type
- Updated tools to handle IPv6 and use AF_* macros

### CODE REVIEW

File: src/ast/irbuilderbpf.cpp [hunk 1]
```diff
@@ -110,6 +110,9 @@ llvm::Type *IRBuilderBPF::GetType(const SizedType &stype)
   {
     switch (stype.size)
     {
+      case 16:
+        ty = getInt128Ty();
```
Comment (dalehamel) [inline src/ast/irbuilderbpf.cpp]: Nice trick

**Comment by dalehamel** [comment]
Making it option is definitely preferred :+1: it  was only required before because it needed to be explicit that only `AF_INET` was supported at the time.

**Comment by mmarchini** [comment]
Updated with:

  - Make first parameter of ntop optional and infer the AF type from the type and size of our address paramenter.
  - Some optimizations:
    - Only allocate what we need on the stack (8 bytes for IPv4, 20 bytes for IPv6)
    - Avoid copying things back and forth on the stack, LLVM doesn't seem smart enough to optimize this.

I'll update docs and tests now.

**Comment by mmarchini** [comment]
For literal integers, ntop will be reduced to essentially two operations:

```
#: (18) r1 = 0x[address][af_type]
#: (7b) *(u64 *)(r10 -x) = r1
```

For example:

```
$ sudo ./src/bpftrace -ve 'BEGIN { printf("%s\n", ntop(0x0100007f));}'
Attaching 1 probe...

Program ID: 405

Bytecode: 
0: (bf) r6 = r1
1: (18) r1 = 0x100007f00000002
3: (7b) *(u64 *)(r10 -8) = r1
4: (b7) r1 = 0
5: (7b) *(u64 *)(r10 -16) = r1
6: (18) r7 = 0xffff8b6d7e5f3100
8: (85) call bpf_get_smp_processor_id#8
9: (bf) r4 = r10
10: (07) r4 += -16
11: (bf) r1 = r6
12: (bf) r2 = r7
13: (bf) r3 = r0
14: (b7) r5 = 16
15: (85) call bpf_perf_event_output#25
16: (b7) r0 = 0
17: (95) exit
processed 16 insns, stack depth 16

Attaching BEGIN
Running...
127.0.0.1
^C

```

Not particularly useful, because we'll fetch the address using probe_read on most cases. But I'm happy with this result :)

**Comment by brendangregg** [comment]
thanks!

**Comment by ajor** [comment]
This patch has broken some stuff related to reading in structs.
e.g. if you try running the test in `tests/codegen/struct_save_nested.cpp` on a real program, it will refuse to load now.

**Comment by ajor** [comment]
Fixed in #579

**Pull request merged**